File: 70-climbing-stairs.py
class Solution:
    def climbStairs(self, n: int) -> int:
        return self.climbStairsMemo(n, {})
        
    def climbStairsMemo(self, n: int, memo) -> int:
        if (n in memo): 
            print(memo)
            return memo[n]

        if (n < 0): return 0;
        if (n == 0 or n == 1): return 1

        memo[n] = self.climbStairsMemo(n - 1, memo) + self.climbStairsMemo(n - 2, memo)
        return memo[n]

File: leetcode/70-climbing-stairs.cpp
#include <vector>
class Solution {
 public:
  int climbStairs(const int n) {
    std::vector<int> dp(n + 1);
    dp[1] = 1;
    dp[0] = 0;
    for (auto i = 2; i <= n; ++i) dp[i] = dp[i - 1] + dp[i - 2];
    return dp[n];
  }
};
